math/minuit2/inc/Minuit2/NumericalDerivator.h

```diff
@@ -0,0 +1,114 @@
+
+   void SetupDifferentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *cx,
+                           const std::vector<ROOT::Fit::ParameterSettings> &parameters);
+   std::vector<DerivatorElement> Differentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                                               const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                                               const std::vector<DerivatorElement> &previous_gradient);
+
+   DerivatorElement PartialDerivative(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                                      const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                                      unsigned int i_component, DerivatorElement previous);
+   DerivatorElement FastPartialDerivative(const ROOT::Math::IBaseFunctionMultiDim *function,
+                                          const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                                          unsigned int i_component, const DerivatorElement &previous);
+   DerivatorElement operator()(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                               const std::vector<ROOT::Fit::ParameterSettings> &parameters, unsigned int i_component,
+                               const DerivatorElement &previous);
+
+   double GetValue() const { return fVal; }
+   void SetStepTolerance(double value);
+   void SetGradTolerance(double value);
+   void SetNCycles(unsigned int value);
+   void SetErrorLevel(double value);
+
+   double Int2ext(const ROOT::Fit::ParameterSettings &parameter, double val) const;
+   double Ext2int(const ROOT::Fit::ParameterSettings &parameter, double val) const;
+   double DInt2Ext(const ROOT::Fit::ParameterSettings &parameter, double val) const;
+
+   void SetInitialGradient(const ROOT::Math::IBaseFunctionMultiDim *function,
+                           const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                           std::vector<DerivatorElement> &gradient);
+
+private:
+   double fStepTolerance = 0.5;
+   double fGradTolerance = 0.1;
+   unsigned int fNCycles = 2;
+   double fUp = 1;
+   double fVal = 0;
+
+   std::vector<double> fVx, fVxExternal;
+   double fDfmin;
+   double fVrysml;
+
+   // MODIFIED: Minuit2 determines machine precision in a slightly different way than
+   // std::numeric_limits<double>::epsilon()). We go with the Minuit2 one.
+   ROOT::Minuit2::MnMachinePrecision fPrecision;
+
+   ROOT::Minuit2::SinParameterTransformation fDoubleLimTrafo;
+   ROOT::Minuit2::SqrtUpParameterTransformation fUpperLimTrafo;
+   ROOT::Minuit2::SqrtLowParameterTransformation fLowerLimTrafo;
+
+private:
+   bool fAlwaysExactlyMimicMinuit2;
+
+public:
+   bool AlwaysExactlyMimicMinuit2() const;
+   void SetAlwaysExactlyMimicMinuit2(bool flag);
+
+private:
+   std::vector<double> fVxFValCache;
```

Comment:
I grouped members according to visibility and size, so the class instances will use less space in RAM.
```suggestion
   bool AlwaysExactlyMimicMinuit2() const;
   void SetAlwaysExactlyMimicMinuit2(bool flag);
   
private:
   double fStepTolerance = 0.5;
   double fGradTolerance = 0.1;
   double fUp = 1;
   double fVal = 0;

   std::vector<double> fVx, fVxExternal;
   double fDfmin;
   double fVrysml;

   // MODIFIED: Minuit2 determines machine precision in a slightly different way than
   // std::numeric_limits<double>::epsilon()). We go with the Minuit2 one.
   ROOT::Minuit2::MnMachinePrecision fPrecision;

   ROOT::Minuit2::SinParameterTransformation fDoubleLimTrafo;
   ROOT::Minuit2::SqrtUpParameterTransformation fUpperLimTrafo;
   ROOT::Minuit2::SqrtLowParameterTransformation fLowerLimTrafo;

   unsigned int fNCycles = 2;
   bool fAlwaysExactlyMimicMinuit2;

   std::vector<double> fVxFValCache;
```

Reply:
I would also put the two vectors in separate lines to make the code more readable (it's easy to miss there are actually two `std::vector` data members when you write `std::vector<double> fVx, fVxExternal;`):
```
std::vector<double> fVx;
std::vector<double> fVxExternal;
```

Reply:
Three, actually! Suggestion updated! 👍